### PR TITLE
feat: allow multiple languages for compilers

### PR DESCRIPTION
**Pull request by klkvr**
This changes `Compiler` trait to also have associated `Language` type which is supposed to be an enum with languages supported by the compiler. Also `CompilerVersionManager` trait is removed and compiler now should contain full logic for version resolution.

For version management we have to keep track of all different languages and versions to compile them with, thus I've rewritten some of the graph logic to use `HashMap<C::Language, HashMap<Version, Sources>>` during compilation.

Now, instead of representing a binary wrapper, `Compiler` is more like binary manager which resolves and installs binaries dynamically depending on requested `language` and `version`.

It's mostly ready, I just need to add logic for compilation with locked compiler version.

### CODE REVIEW

File: src/compilers/solc/mod.rs [hunk 1]
```diff
@@ -1,39 +1,55 @@
...
+#[derive(Debug, Clone, Default)]
+#[non_exhaustive]
+pub struct SolcRegistry;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[non_exhaustive]
+pub enum SolcLanguages {
+    Solidity,
+    Yul,
+}
```
Comment (mattsse) [inline src/compilers/solc/mod.rs:33]: this needs some docs

File: src/compilers/solc/mod.rs [hunk 2]
```diff
@@ -1,39 +1,55 @@
...
 };
 
-impl Compiler for Solc {
+#[derive(Debug, Clone, Default)]
+#[non_exhaustive]
+pub struct SolcRegistry;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[non_exhaustive]
+pub enum SolcLanguages {
```
Comment (mattsse) [inline src/compilers/solc/mod.rs:30]: nit: 
```suggestion
pub enum SolcLanguage {
```

File: src/compilers/mod.rs [hunk 3]
```diff
@@ -136,45 +206,31 @@ impl<E> Default for CompilerOutput<E> {
     }
 }
 
+pub trait Language: Hash + Eq + Clone + Debug {
```
Comment (mattsse) [inline src/compilers/mod.rs:209]: should this also be Display and I think this should be `'static`

File: src/compilers/solc/mod.rs [hunk 4]
```diff
@@ -44,85 +60,80 @@ impl Compiler for Solc {
...
+        all_versions
     }
+}
 
-    fn with_include_paths(mut self, include_paths: BTreeSet<PathBuf>) -> Self {
-        self.include_paths = include_paths;
-        self
-    }
+#[derive(Debug, Clone, Serialize)]
+pub struct SolcVerionedInput {
```
Comment (mattsse) [inline src/compilers/solc/mod.rs:84]: typo

File: src/compilers/solc/mod.rs [hunk 5]
```diff
@@ -1,39 +1,55 @@
...
+    Solidity,
+    Yul,
+}
+
+impl Language for SolcLanguages {
     const FILE_EXTENSIONS: &'static [&'static str] = SOLC_EXTENSIONS;
+}
 
-    type Input = SolcInput;
+impl Compiler for SolcRegistry {
```
Comment (mattsse) [inline src/compilers/solc/mod.rs:39]: hmm, it's a bit weird that we implement compiler on registry, maybe we can rename this? so it's more obvious that this is compiler?

**Review by mattsse — CHANGES_REQUESTED** [review]
this makes a lot of sense and simplifies things a lot.

I only have doc nits

File: src/compilers/solc/mod.rs (hunk 5 above)
Comment (klkvr) [reply src/compilers/solc/mod.rs:39]: yeah, not sure how to name it

perhaps `MultiSolc` or just `SolcCompiler`?

Comment (mattsse) [reply src/compilers/solc/mod.rs:39]: I think SolcCompiler fits

Comment (DaniPopes) [reply src/compilers/solc/mod.rs:39]: Sol Compiler Compiler

**Review by mattsse — APPROVED** [review]
gg

**Pull request merged**